bokeh/io.py

```diff
@@ -424,6 +430,38 @@ def push(session_id=None, url=None, app_path=None, document=None, state=None, io
     _push_to_server(session_id=session_id, url=url, app_path=app_path,
                     document=document, io_loop=io_loop)
 
+def push_notebook(document=None, state=None):
+    ''' Update the notebook with the data for the current document.
+
+    Args:
+
+        document (Document, optional) : A :class:`bokeh.document.Document` to use
+
+        state (State, optional) : A state to use for any output_server() configuration of session or url
+
+    Returns:
+        None
+
+    '''
+    if state is None:
+        state = _state
+
+    if not document:
+        document = state.document
+
+    if not document:
+        warnings.warn("No document to push")
+        return
+
+    if not state.last_comms:
+        warnings.warn("Cannot find a last shown plot to update")
+        return
+
+    # TODO prepare actual patch
```

Comment:
Some ideas:

1) each time we send the doc, save its JSON. When we want to patch it, compute new json and diff vs old json. Will have to port the JSON diff algorithm we have over in document.coffee to Python. Downside: have to port the diff code. 

2) each time we send the document, connect to on_change and start recording all events. when we want to patch, replay them all as patch-doc. Downside: event record could grow infinitely in theory.

3) resend entire doc each time. Downside: sends entire doc to change one property.

There could be issues when coffee makes changes during document rendering; those changes won't be synced back. So the events may not always apply correctly, depending on what they are, because the Python side won't have an accurate copy of the document.